### DISCUSSION TITLE
Rename audio source for PWAs

### DISCUSSION THREAD

**Discussion by Zhincore** [filips123/PWAsForFirefox#445] [Ideas]
Hello,
would it be possible to rename the pwa in terms of pulseaudio/pipewire, so that the audio source is not called "Firefox" like the base firefox? 
I use EasyEffects and I want to apply it's effects to YouTube Music PWA but not the base Firefox process, which is possible, but EE remembers "Firefox" and after reboot applies itself to both processes since they're called the same.

![Screenshot of the two processes in EasyEffects](https://github.com/filips123/PWAsForFirefox/assets/21307300/9491cb15-c89b-401c-9da6-165677195276)
Screenshot of the two processes in EasyEffects, top is installed PWA and bottom is the base app, basically indistinguishable.

I am not sure how independent the PWA is so I don't know how possible this is. I think there is a way to rename during runtime, but maybe I am wrong.
Thank you

**Comment by filips123** [comment]
How does EasyEffects/PulseAudio/PipeWire distinguish between programs?

PWAsForFirefox already sets a window name and class for each PWA, which is used to create separate desktop/toolbar icons, but it seems EasyEffects/PulseAudio uses some different mechanism to detect app names.

I found [this](https://www.freedesktop.org/wiki/Software/PulseAudio/Documentation/Developer/Clients/ApplicationProperties/). However, setting the application name and icon with `g_set_application_name` and `gtk_window_set_default_icon_name` won't work because that will be shared across all open PWAs in the same profile. Maybe it's possible to do something with `PULSE_PROP_application.name` and other environment variables, but I'm not sure yet. I'll probably look into this in the future.

I think it's possible to set environment variable using `Services.env.set`, but I haven't yet tested this if it actually works with PulseAudio.

**Reply by Zhincore** [reply]
Hi, thank you for the quick response!
I am not sure how it works, but my google "research" agrees that `PULSE_PROP_application.name` might be the solution! 
I blindly tried putting `Services.env.set("PULSE_PROP_application.name", "FirefoxPWA")` in various places in various scripts in my firefoxpwa profile and runtime folder, but had no luck. 
I think it's possible that cubeb, the library firefox uses for multi-platform audio, might be overriding the env forcefully.
So I tried rebranding the firefox instance but I didn't even manage to do that, so I don't know... I thought using the `-app` command line argument I may rebrand it, but that doesn't seem to be the case...

**Reply by Zhincore** [reply]
By the way, is it necessary to have the option to have PWAs on the same profile? Would something break if you forced the PWAs to be "isolated"? Just curious

**Reply by filips123** [reply]
> I blindly tried putting `Services.env.set("PULSE_PROP_application.name", "FirefoxPWA")` in various places in various scripts in my firefoxpwa profile and runtime folder, but had no luck.

You can try to put it [around here](https://github.com/filips123/PWAsForFirefox/blob/main/native/userchrome/profile/chrome/pwa/boot.jsm#L194) (`/usr/share/firefoxpwa/userchrome/profile/chrome/pwa/boot.jsm`). Then, you need to enable "Always patch runtime and profile" in the extension settings and launch your PWA.

If that doesn't work, maybe environment variables set with `Services.env.set` don't affect all processes they would need to. In this case, implementing this will probably be harder.

Another thing you can try is to launch PWA from terminal with:

```shell
PULSE_PROP_application.name=FirefoxPWA firefoxpwa site launch SITE-ID
```

Maybe at least this will properly set environment variable.

I will also check that in the future, but I don't have time currently.

> By the way, is it necessary to have the option to have PWAs on the same profile? Would something break if you forced the PWAs to be "isolated"? Just curious

Well, currently on Linux (and macOS), it's recommended to have separate/isolated profiles because of some other issues (#322). But I want to fix those issues eventually so PWAs work properly even if they are in the same profile, as that is often more convenient for users. So, implementing a solution here that only works when PWAs are separate won't be that good as I would then need to change it in the future.

**Reply by Zhincore** [reply]
Shell doesn't like the name of the variable so I tried `env PULSE_PROP_application.name="FirefoxPWA" firefoxpwa site launch ...` instead, but the name still hasn't changed.

Thank you for your efforts! I will be looking forward for if you find a solution in the future.

**Reply by filips123** [reply]
I researched Firefox code a bit and I found that Firefox sets the app name based on the [`brandShortName`](https://searchfox.org/mozilla-central/source/dom/media/CubebUtils.cpp#387-409) (on Linux) or [`brandFullName`](https://searchfox.org/mozilla-central/source/widget/windows/AudioSession.cpp#150-154) (on Windows) property from `chrome://branding/locale/brand.properties`. It seems there is no other way to directly overwrite the audio app name. And I still don't know what is used on macOS.

I don't know if it's possible to overwrite these brand properties. And even if it's possible, I don't know if that can cause other unintended consequences, as brand properties are also used elsewhere in Firefox UI and internally.

I also still don't know why the `PULSE_PROP_application.name` environment variable doesn't work. Maybe it's because Firefox itself sets the name, so the audio library then doesn't look up the name in the environment variable. Setting the `PULSE_PROP_application.icon_name` environment variable worked (the icon showed in EasyEffects), but setting just the icon isn't that useful, and it only works for one PWA per profile. Setting the `PULSE_PROP_application.id` did something, but it caused some problems in EasyEffects (exclude doesn't work anymore), and it also only works for one PWA per profile.

So, unfortunately, I think it's not easily possible to do this. I would probably have to modify Firefox to add internal JS API to change the audio name, icon and ID. Even with such API, it may not be possible to customize properties if multiple PWAs are installed in the same profile, but it could at least work for one PWA per profile. I may try to create this API in the future, but I'm not sure when.

Another option could be to check why `PULSE_PROP_application.name` doesn't work and why `PULSE_PROP_application.id` causes problems in EasyEffects, and try to fix that. However, I don't even know where to begin and which library is supposed to read them...
